classes/class-p4-search.php

```diff
@@ -116,9 +123,132 @@ protected function initialize() {
+						unset( $meta['sm_cloud'] );
+					}
+					if ( !empty( $meta['_wp_attachment_image_alt'] ) ) {
+						$meta['_wp_attachment_image_alt'] = [ $meta['_wp_attachment_image_alt'][0] ];
+					}
+					return $meta;
+				},
+				20,
+				2
+			);
+			// Switch the language to the language of the current post while doing a sync of a post.
+			// This is needed because otherwise WPML black magic will translate some parts of synced data for this post
+			// into the language of the current admin, causing search results in the wrong language. As these syncs are
+			// queued and then executed in random wp admin requests, this language is pretty much random.
+			// It's safe to switch the language in this filter, as the ES queue is only processed at the very end of the
+			// request, and calling `switch_lang` has no effect outside the request.
+			// We need to abuse the `ep_ignore_invalid_dates` filter as it's the only one available to switch the
+			// language in time that has the post available.
+			add_filter(
+				'ep_ignore_invalid_dates',
+				function ( $ignore, $post_id, $post ) {
+					/**
+					 * @var SitePress
+					 */
+					global $sitepress;
+					if ( $sitepress ) {
+						$lang_code = ( new WPML_Post_Element( $post->ID, $sitepress ) )->get_language_code();
+						$sitepress->switch_lang( $lang_code );
+					}
+				},
+				20,
+				3
+			);
+			// Specify which fields to fetch so that we don't need to fetch the post content.
+			add_filter(
+				'ep_formatted_args',
+				function ( $formatted_args, $args ) {
+					$formatted_args['_source'] = [
+						'post_id',
+						'ID',
+						'post_author',
+						'post_date',
+						'post_date_gmt',
+						'post_title',
+						'post_excerpt',
+						'post_name',
+						'post_modified',
+						'post_modified_gmt',
+						'post_content',
+						'post_parent',
+						'post_type',
+						'post_mime_type',
+						'permalink',
+						'post_slug',
+						'terms',
+						'date_terms',
+						'comment_count',
+						'comment_status',
+						'guid',
+						'post_lang',
+					];
+					return $formatted_args;
+				},
+				10,
+				2
+			);
+			// Make it return the post guid as that's where we put the external url for the archive post type.
+			add_filter(
+				'ep_search_post_return_args',
+				function ( $args ) {
+					$args[] = 'guid';
+					$args[] = 'terms';
+
+					return $args;
+				}
+			);
+			// Because of how the search is currently set up (using admin-ajax) this ElasticPress filter was not being
+			// applied for subsequent page loads, only for the initial one.
+			if ( ( ! isset( $_GET['orderby'] ) || '_score' === $_GET['orderby'] ) && wp_doing_ajax() ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+				add_filter(
+					'ep_formatted_args',
+					[ new \ElasticPress\Feature\Search\Search(), 'weight_recent' ],
+					10,
+					2
+				);
+			}
+			add_filter( 'posts_where', [ self::class, 'edit_search_mime_types' ] );
+			remove_filter(
```

Comment:
This was causing the issue where documents show up for other types. If document search is enabled in ElasticPress this filter added this to the query, after our code has determined the post_types. We can safely remove this filter as we set attachment as a type ourselves.